packages/vitest/src/runtime/entry.ts

```diff
@@ -22,45 +21,29 @@ export async function run(files: string[], config: ResolvedConfig): Promise<void
   const filesWithEnv = await Promise.all(files.map(async (file) => {
     const code = await fs.readFile(file, 'utf-8')
     const env = code.match(/@(?:vitest|jest)-environment\s+?([\w-]+)\b/)?.[1] || config.environment || 'node'
+    const envOptions = JSON.parse(code.match(/@(?:vitest|jest)-environment-options\s+?(.+)/)?.[1] || 'null')
     return {
       file,
       env: env as VitestEnvironment,
+      envOptions: envOptions as EnvironmentOptions | undefined,
     }
   }))
 
-  const filesByEnv = filesWithEnv.reduce((acc, { file, env }) => {
```

Comment:
Just a hint. It is required to group files by env for `--no-threads`. Now, you need to group them by env AND options: https://vitest.dev/config/#environment

> If you are running Vitest with [--no-threads](https://vitest.dev/config/#threads) flag, your tests will be run in this order: node, jsdom, happy-dom, edge-runtime, custom environments. Meaning, that every test with the same environment is grouped together, but is still running sequentially.

Reply:
Out of curiosity, is there a specific reason we need to sort by configs, and not just envs?

Reply:
Well, if environment has different config, it's a different environment, isn't it? If you only group by environments, then only one config will be applied.